ngraph/core/reference/include/ngraph/runtime/reference/adaptive_avg_pool.hpp

```diff
@@ -0,0 +1,177 @@
+                        size_t d_end = window_end(i, d_in, d_out);
+                        for (size_t j = 0; j < h_out; j++)
+                        {
+                            size_t h_start = window_start(j, h_in, h_out);
+                            size_t h_end = window_end(j, h_in, h_out);
+                            for (size_t k = 0; k < w_out; k++)
+                            {
+                                size_t w_start = window_start(k, w_in, w_out);
+                                size_t w_end = window_end(k, w_in, w_out);
+                                T result = 0;
+                                for (size_t n = d_start; n < d_end; n++)
+                                {
+                                    for (size_t m = h_start; m < h_end; m++)
+                                    {
+                                        auto pos = arg + n * h_in * w_in + m * w_in;
+                                        result =
+                                            std::accumulate(pos + w_start, pos + w_end, result);
+                                    }
+                                }
+                                out[i * h_out * w_out + j * w_out + k] = avg_div(
+                                    result,
+                                    (d_end - d_start) * (w_end - w_start) * (h_end - h_start));
+                            }
+                        }
+                    }
+                }
+            } // namespace adaptive_pool
+            template <typename T>
+            void adaptive_avg_pool(const T* arg,
+                                   T* out,
+                                   const Shape& arg_shape,
+                                   const Shape& out_shape)
+            {
+                NGRAPH_CHECK(arg_shape.size() == out_shape.size() && 2 < arg_shape.size() &&
+                                 arg_shape.size() < 6,
+                             "AdaptiveAvgPool supports only 3D, 4D and 5D input shape");
+                size_t channel_size = 1;
+                for (size_t i = 2; i < arg_shape.size(); i++)
+                {
+                    channel_size *= arg_shape[i];
+                }
+                size_t batch_size = arg_shape[1] * channel_size;
+                size_t out_channel_size = 1;
+                for (size_t i = 2; i < out_shape.size(); i++)
+                {
+                    out_channel_size *= out_shape[i];
+                }
+                size_t out_batch_size = arg_shape[1] * out_channel_size;
+                for (size_t b = 0; b < arg_shape[0]; b++)
+                {
+                    for (size_t c = 0; c < arg_shape[1]; c++)
+                    {
+                        auto arg_pos = arg + b * batch_size + c * channel_size;
+                        auto out_pos = out + b * out_batch_size + c * out_channel_size;
+                        if (arg_shape.size() == 3 && out_shape.size() == 3)
```

Comment:
I think you can simplify this check
```suggestion
                        if (arg_shape.size() == 3)
```
Because you already have a check: `NGRAPH_CHECK(arg_shape.size() == out_shape.size())`